examples/pwa.js

```diff
@@ -0,0 +1,321 @@
+      } else if (isMeta && n.hasAttribute('charset')) {
+        // Ignore.
+      } else if (isMeta && name == 'viewport') {
+        // Ignore.
+      } else if (isLink && rel == 'canonical') {
+        this.canonicalUrl_ = n.getAttribute('href');
+        log('- canonical: ', this.canonicalUrl_);
+      } else if (isLink && rel == 'stylesheet') {
+        this.stylesheets_.push(n.getAttribute('href'));
+        log('- stylesheet: ', this.stylesheets_[this.stylesheets_.length - 1]);
+      } else if (n.tagName == 'STYLE') {
+        if (n.hasAttribute('amp-boilerplate')) {
+          // Ignore.
+          log('- ignored embedded style: ', n);
+        } else {
+          log('- embedded style: ', n);
+          this.shadowRoot_.appendChild(this.win.document.importNode(n, true));
+        }
+      } else if (n.tagName == 'SCRIPT') {
+        if (n.hasAttribute('src')) {
+          log('- src script: ', n);
+          this.scripts_.push(n);
+        } else {
+          log('- non-src script: ', n);
+          this.shadowRoot_.appendChild(this.win.document.importNode(n, true));
+        }
+      } else if (n.tagName == 'NOSCRIPT') {
+        // Ignore.
+      } else {
+        log('- UNKNOWN head element:', n);
+      }
+    }
+
+    this.mergeHead_();
+
+    // Body
+    this.shadowRoot_.appendChild(this.win.document.importNode(doc.body, true));
+  }
+
+  mergeHead_() {
+    const doc = this.win.document;
+
+    // Title.
+    doc.title = this.title_ || '';
+    log('SET title: ', doc.title);
+
+    // Stylesheets.
+    this.stylesheets_.forEach(stylesheet => {
+      const href = this.resolveUrl_(stylesheet);
+      const exists = doc.querySelector('link[href="' + href + '"]');
+      if (exists) {
+        log('- stylesheet already exists: ', href);
+      } else {
+        const el = doc.createElement('link');
+        el.setAttribute('rel', 'stylesheet');
+        el.setAttribute('type', 'text/css');
+        el.setAttribute('href', href);
+        doc.head.appendChild(el);
+        log('- stylesheet added: ', href, el);
+      }
+    });
+
+    // Scripts.
+    this.scripts_.forEach(script => {
+      // XXX: stub elements, reg templates
+      const customElement = script.getAttribute('custom-element');
+      const customTemplate = script.getAttribute('custom-template');
+      const src = this.resolveUrl_(script.getAttribute('src'));
+      log('script: ', customElement, customTemplate, script.getAttribute('src'), src);
+      const existsExpr =
+          customElement ? '[custom-element="' + customElement + '"]' :
+          customTemplate ? '[custom-template="' + customTemplate + '"]' :
+          // TODO: version control!
+          '[src="' + src + '"]';
+      const exists = doc.querySelector('script' + existsExpr);
+      if (exists) {
+        log('- script already exists: ', customElement, customTemplate, src);
+      } else {
+        const el = doc.createElement('script');
+        el.setAttribute('async', '');
```

Comment:
Yeah. Just wanted it for debugging. 

`data-` - it's fine. But ref our conversation re: injecting the extensions via runtime.